2019/day2/day2.go

```diff
@@ -0,0 +1,102 @@
+package day2
+
+import (
+	"errors"
+	"fmt"
+)
+
+const (
+	opcodeAdd       = 1
+	opcodeMultiply  = 2
+	opcodeTerminate = 99
+)
+
+var opcodeInstructionCount = map[int]int{
+	opcodeAdd:       4,
+	opcodeMultiply:  4,
+	opcodeTerminate: 1,
+}
+
+var (
+	errInvalidOpcode = errors.New("invalid opcode")
+	errInvalidLen    = errors.New("invalid length")
+	errNounRange     = errors.New("noun out of [0,99] range")
+	errVerbRange     = errors.New("verb out of [0,99] range")
+)
+
+func Part1(input []int) (int, error) {
+	const (
+		noun = 12
+		verb = 2
+	)
+	if err := restore(noun, verb, input); err != nil {
+		return -1, err
+	}
+	return IntcodeProgram(input)
+}
+
+func restore(noun, verb int, input []int) error {
+	const minInputLen = 3
+	if len(input) < minInputLen {
+		return fmt.Errorf("%w: minimum length is %d, got %d", errInvalidLen, minInputLen, len(input))
+	}
+	if noun < 0 || noun > 99 {
+		return fmt.Errorf("%w: %d", errNounRange, noun)
+	}
+	if verb < 0 || verb > 99 {
+		return fmt.Errorf("%w: %d", errVerbRange, verb)
+	}
+	input[1] = noun
+	input[2] = verb
+	return nil
+}
+
+func IntcodeProgram(input []int) (int, error) {
+	var programCounter int
+	for programCounter < len(input) {
+		op := input[programCounter]
+		switch {
```

Comment:
You can use `switch op {` and then you don't need the equals part below.

Reply:
changing this line to `switch op {` made it so that `opCodeAdd` and `opCodeMultiply` need to be their own case which seems fine. `opCodeAdd` logic is just different enough from `opCodeMultiply` that it seems ok to break them out into their own case. The only part I slightly don't like is that I have to duplicate the code accessing items from the input array, [ref](https://github.com/JessicaGreben/adventofcode/pull/1/files#diff-c7567595d2c93ccef1601fe41ac27bda86ab74cb445453c8350c815a96e62c76R61).

Reply:
If you want to handle multiple things in the same case, you can use:
```
switch op := opCode(input[programCounter]); op {
case opCodeAdd, opCodeMultiply:
     ...
```

However, I would keep this version, because handling them individually seems clearer.

Reply:
ah i see. i couldn't figure out that syntax. which do you think is better? keeping them separate or putting them in one case statement?

Reply:
I think either is fine. I would probably keep it separate and write helper func for reading the two arguments.

Reply:
I added a helper func to read the arguments. Thoughts?

Reply:
Looks good, but I would be more specific with regards to the func name. e.g. `Read_2In_1Out`,`ReadI2O1` or something -- just to make it obvious what you expect the code to be.